### ISSUE TITLE
[Python] Conversion to/from numpy 2.0+ new StringDType

### ISSUE THREAD

**Issue by jorisvandenbossche** [apache/arrow#42018]
With the upcoming numpy 2.0 release, they introduce a new variable-length string dtype, [StringDType](https://numpy.org/devdocs/reference/routines.dtypes.html#numpy.dtypes.StringDType) (https://numpy.org/devdocs/release/2.0.0-notes.html#highlights). 

Currently trying to convert that obviously gives an error:

```python
>>> arr = np.array(["some", "strings"], dtype=np.dtypes.StringDType())
>>> arr
array(['some', 'strings'], dtype=StringDType())
>>> pa.array(arr)
...
ArrowNotImplementedError: Unsupported numpy type 2056
```

But ideally we should support this dtype both for conversion from numpy -> arrow, as in the arrow -> numpy conversion (although here it should probably be opt-in and by default still return object dtype)

Numpy provides a C API to access the individual string elements: https://numpy.org/devdocs/reference/c-api/strings.html

**Comment by alippai** [comment]
I believe this is not stale, numpy is continuously improving the StringDType support.

@jorisvandenbossche would this be implemented on the C++ side or cython/python?

**Comment by thisisnic** [comment]
This issue has been closed because there has been no activity for the 14 days since the stale warning. Feel free to reopen if this has been closed in error.

**Comment by alippai** [comment]
@thisisnic there was clear activity (both comment and pending MR). Is the script broken?

**Comment by thisisnic** [comment]
Doing it manually due to GitHub API limits on the CI job, and it's missed the logic for where the label has been left but someone has commented. Will reopen!

**Comment by alippai** [comment]
Thanks for keeping the issue queue tidy

**Comment by rok** [comment]
cc @ngoldbaum

**Comment by alippai** [comment]
@rok this waits for numpy minimum version bump only. There was a debate about the different settings and requiring 2.0+ is the only version simple enough for everyone

**Comment by rok** [comment]
@alippai that was the consensus here at EuroPython sprints too. NumPy is likely going to be set to 2.0+ with the next release and there is some interest in solving this issue. Did you want to work on this yourself or would you like someone else takes it on?

**Comment by ngoldbaum** [comment]
Hi @alippai I'd like to help to make sure support for StringDType lands. Is there any way I can help? Are you still interested in pushing this forward once the minimum supported NumPy release is NumPy 2.0?

**Comment by ngoldbaum** [comment]
Ah I see from https://github.com/apache/arrow/pull/50769 that you're planning to file a followup for that. Let me know if there's any way I can help out.

**Comment by alippai** [comment]
@ngoldbaum yes, after bumping the numpy & pandas requirement I'll give it another try. I'd be more than happy to accept your review and guidance. I guess it's the best if I submit a PR and we can continue from there.

**Comment by alippai** [comment]
@ngoldbaum thanks for the guide and review on Numpy 2.0+. I’ll continue with this over the weekend.
